# Excel to Data Frame Routines

<b>Routine 1: Concatenating all the data from all the tabs of an excel file into one dataframe - Horizontally. </b>

For, example we have multiple excel files with multiple tabs and columns in it, the structuring of each excel file can be similar. We are trying to extract all the data from these excel files into a Pandas data frame, all the tabs will be concatenated horizontally (also their data). The Last column would specify the excel file name from which the data has been extracted. 

While extracting the data we're storing each excel file in a list, i.e., each of the pandas data frame is stored in a list. If we start looking into the elements of the list we will get a Pandas data frame corresponding to each excel file.

Below is the code for the above mentioned procedure:

In [1]:
import os
import xlrd
import numpy as np
import pandas as pd

In [2]:
def excel_to_df_1():
    '''This routine takes multiple excel files and converts them into list of data frames, 
    concatenating the data horizontally'''
    
    #Assigning the path to the folder variable. Please assign the correct path to the folder where your files are saved.
    folder = r'C:/Users/DELL/Desktop/Activate/AI Training Package/Surveys/Word Parsing/Results/'
    
    #Getting the list of files from the assigned path
    excel_files = [file for file in os.listdir(folder)]
    
    #This list contains all the pandas data frame corresponding to each excel file
    list_of_dfs = []
    
    for file in excel_files :
        
        #pd.concat concatenates all the tabs (as we provide sheet_name = None) 
        df = pd.concat(pd.read_excel(folder + "\\" + file, sheet_name=None), ignore_index=True)
        
        #This line adds an column named 'excelfile_name' at the end of each data frame and it adds the file name as the value
        df['excelfile_name'] = file.split('.')[0]
        
        #Lastly we append each dataframe into the list
        list_of_dfs.append(df)
        
    return list_of_dfs

In [3]:
list_of_dfs = excel_to_df_1()
print("Type of the elements in the list: ", type(list_of_dfs[0]))

Type of the elements in the list:  <class 'pandas.core.frame.DataFrame'>


In [4]:
list_of_dfs[0].shape

(1392, 145)

In [5]:
list_of_dfs[0].head(100)

Respondent ID  Collector ID          Start Date            End Date  \
0             NaN           NaN                 NaT                 NaT   
1    1.793692e+09    26147937.0 2012-04-11 17:46:11 2012-04-11 18:10:36   
2    1.793354e+09    26147937.0 2012-04-11 14:31:08 2012-04-11 14:40:43   
3    1.791946e+09    26147937.0 2012-04-10 16:54:05 2012-04-10 17:25:21   
4    1.791862e+09    26147937.0 2012-04-10 16:04:49 2012-04-10 16:39:17   
..            ...           ...                 ...                 ...   
95   1.791158e+09           NaN                 NaT                 NaT   
96   1.791158e+09           NaN                 NaT                 NaT   
97   1.791158e+09           NaN                 NaT                 NaT   
98   1.791158e+09           NaN                 NaT                 NaT   
99   1.791158e+09           NaN                 NaT                 NaT   

       IP Address  Email Address  First Name  Last Name  Custom Data 1  \
0             NaN            NaN         NaN        NaN            NaN   
1   68.15.196.223            NaN         NaN        NaN            NaN   
2   68.15.196.223            NaN         NaN        NaN            NaN   
3   68.15.196.223            NaN         NaN        NaN            NaN   
4   68.15.196.223            NaN         NaN        NaN            NaN   
..            ...            ...         ...        ...            ...   
95            NaN            NaN         NaN        NaN            NaN   
96            NaN            NaN         NaN        NaN            NaN   
97            NaN            NaN         NaN        NaN            NaN   
98            NaN            NaN         NaN        NaN            NaN   
99            NaN            NaN         NaN        NaN            NaN   

   Please indicate at what level you believe that Astellas establishes/practices the following attributes regarding STRUCTURE.  Although you may see evidence at times to support multiple ratings, select the option you feel most closely reflects the organization's overall functioning.   \
0   Clear roles and responsibilities of all employees                                                                                                                                                                                                                                           
1                                                   3                                                                                                                                                                                                                                           
2                                                   2                                                                                                                                                                                                                                           
3                                                   3                                                                                                                                                                                                                                           
4                                                   3                                                                                                                                                                                                                                           
..                                                ...                                                                                                                                                                                                                                           
95                                                NaN                                                                                                                                          

This shows that our first excel file has total of 145 columns in all the tabs and 1392 rows. The below code shows the similar information.

In [6]:
list_of_dfs[0].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1392 entries, 0 to 1391
Columns: 145 entries, Respondent ID to excelfile_name
dtypes: datetime64[ns](2), float64(16), object(127)
memory usage: 1.5+ MB


<b>Routine 2: Concatenating all the data from all the tabs of an excel file into one dataframe - Horizontally with tab names. </b>

Here, we are trying to extract all the data from these excel files into a Pandas data frame, all the tabs will be concatenated horizontally (also their data), with the first column in the data frame showing the tab name. The Last column would specify the excel file name from which the data has been extracted. 

While extracting the data we're storing each excel file in a list, i.e., each of the pandas data frame is stored in a list. If we start looking into the elements of the list we will get a Pandas data frame corresponding to each excel file.

Below is the code for the above mentioned procedure:

In [7]:
def excel_to_df_2():
    '''This routine takes multiple excel files and converts them into list of data frames, concatenating the data vertically 
    with tab name as the first column'''
    
    #Assigning the path to the folder variable
    folder = 'C:/Users/DELL/Desktop/Activate/AI Training Package/Surveys/Word Parsing/Results/'
    
    #Getting the list of files from the assigned path
    excel_files = [file for file in os.listdir(folder)]
    
    #List to capture all the data frames
    list_of_dfs1 = []
    
    for file in excel_files :
        
        #Concatenating the data from the excel file
        df = pd.concat(pd.read_excel(folder + "\\" + file, sheet_name=None))
        
        #Getting the index value and assigning it to the first column
        df.index = df.index.get_level_values(0)
        df.reset_index().rename({'index':'Tab'}, axis=1)
        
        #Assigning the last column with the excel file name
        df['excelfile_name'] = file.split('.')[0]
        
        list_of_dfs1.append(df)
    
    return list_of_dfs1

In [8]:
list_of_dfs1 = excel_to_df_2()
print("Type of the elements in the list: ", type(list_of_dfs1[0]))

Type of the elements in the list:  <class 'pandas.core.frame.DataFrame'>


In [9]:
list_of_dfs1[0].shape

(1392, 145)

In [10]:
list_of_dfs1[0].head(100)

Respondent ID  Collector ID          Start Date            End Date  \
Sheet               NaN           NaN                 NaT                 NaT   
Sheet      1.793692e+09    26147937.0 2012-04-11 17:46:11 2012-04-11 18:10:36   
Sheet      1.793354e+09    26147937.0 2012-04-11 14:31:08 2012-04-11 14:40:43   
Sheet      1.791946e+09    26147937.0 2012-04-10 16:54:05 2012-04-10 17:25:21   
Sheet      1.791862e+09    26147937.0 2012-04-10 16:04:49 2012-04-10 16:39:17   
...                 ...           ...                 ...                 ...   
Comments   1.791158e+09           NaN                 NaT                 NaT   
Comments   1.791158e+09           NaN                 NaT                 NaT   
Comments   1.791158e+09           NaN                 NaT                 NaT   
Comments   1.791158e+09           NaN                 NaT                 NaT   
Comments   1.791158e+09           NaN                 NaT                 NaT   

             IP Address  Email Address  First Name  Last Name  Custom Data 1  \
Sheet               NaN            NaN         NaN        NaN            NaN   
Sheet     68.15.196.223            NaN         NaN        NaN            NaN   
Sheet     68.15.196.223            NaN         NaN        NaN            NaN   
Sheet     68.15.196.223            NaN         NaN        NaN            NaN   
Sheet     68.15.196.223            NaN         NaN        NaN            NaN   
...                 ...            ...         ...        ...            ...   
Comments            NaN            NaN         NaN        NaN            NaN   
Comments            NaN            NaN         NaN        NaN            NaN   
Comments            NaN            NaN         NaN        NaN            NaN   
Comments            NaN            NaN         NaN        NaN            NaN   
Comments            NaN            NaN         NaN        NaN            NaN   

         Please indicate at what level you believe that Astellas establishes/practices the following attributes regarding STRUCTURE.  Although you may see evidence at times to support multiple ratings, select the option you feel most closely reflects the organization's overall functioning.   \
Sheet     Clear roles and responsibilities of all employees                                                                                                                                                                                                                                           
Sheet                                                     3                                                                                                                                                                                                                                           
Sheet                                                     2                                                                                                                                                                                                                                           
Sheet                                                     3                                                                                                                                                                                                                                           
Sheet                                                     3                                                                                                                                                                                                                                           
...                                                     ...                                                                                                                                                                                                                                           
Comments   